Host/libeCos/src/net/tcpip/v2_0/doc/getserv.man
NAME
       getservent, getservbyname, getservbyport - get service entry

SYNOPSIS
       #include <network.h>

       struct servent *getservbyname(const char *name, const char *proto);

       struct servent *getservbyport(int port, const char *proto);

DESCRIPTION
       The  getservbyname()  function returns a servent structure
       for the line from /etc/services that matches  the  service
       name using protocol proto.

       The  getservbyport()  function returns a servent structure
       for the line that matches the port port given  in  network
       byte order using protocol proto.

       The servent structure is defined in <netdb.h> as follows:

              struct servent {
                      char    *s_name;        /* official service name */
                      char    **s_aliases;    /* alias list */
                      int     s_port;         /* port number */
                      char    *s_proto;       /* protocol to use */
              }

       The members of the servent structure are:

       s_name The official name of the service.

       s_aliases
              A zero terminated list of alternative names for the
              service.

       s_port The  port  number  for the service given in network
              byte order.

       s_proto
              The name of the protocol to use with this  service.

RETURN VALUE
       The   getservbyname()  and  getservbyport()
       functions return the servent structure, or a NULL  pointer
       if an error occurs.




ChangeLog.txt
/***** DONT REMOVE OR MODIFY THE FOLLOWING 4 LINES! *****/
  "$Revision: 3705 $",
  "$Id: ChangeLog.txt 3705 2010-03-10 01:53:10Z xhchen $",
  "$Build: "__DATE__" "__TIME__ " $"
/********************************************************/

/*
2010 03 10
    1. Add project HelloWorld for libeCos and open libeCos's source.

2010 01 12
    1. Update new HTML.
    2. Fixed the problem of RTSP buffer overflow.
    
2009 12 24
    1. Reset IP trigger event after DHCP renewed.

2009 11 24
    1. Use random delay time before reconnect PPP server.

2009 11 12
    1. Remove USB debug message.
    2. PPP can reconnect on error.

2009 10 04
    1. Fixed VLC/QuickTime can not play RTSP for long time.

2009 10 23
    1. Fixed that cyg_fd_free may access unknown area.
    2. LED use its own lock instead of period task's lock.

2009 10 16
    1. Can output debug message by VCOM.
    2. SetPPP and GetPPP by VCOM.

2009 10 09
    1. Use new ActiveX version 1.0.0.9.

2009 09 17
    1. Can support up to 10 PPP servers.
    2. Win7 can display MP4 and use VCOM. (VCOM still has problems).
    3. More stable ActiveX.
    4. Add timeout for RTSP TCP connection.

2009 08 14
    1. IPCamConfigure can getback password.

2009 08 12
    1. PPP: Fixed the bug of block on fd close.
    2. PPP: Fixed the bug of block on cyg_io_write/splany.

2009 08 04
    1. Add SetPPP.cgi/GetPPP.cgi.
    2. Fixed ActiveX version error.

2009 07 02
    1. Make nupdate workable on XP.

2009 06 05
    1. MJPEG sometimes slow, fixed.

2009 04 27
    1. Make console usb tool and verify script compatible with Rovio 
       v4.02.
    2. Fixed the bug about add/del user and sprintf.

2009 04 24
    1. Fixed problem ActiveX blocks. (HttpSendRequestEx still blocks).
    2. Add the script to change ActiveX version and clsid.

2009 04 21
    1. Apply CP's update for UART test.

2009 04 20
    1. Break USB writing on USB error.
    2. USB console tools:
       a. Fixed the error in percent value of updating firmware.
       b. Return value fo initSerialConnect() is not correct.
    3. USBWrite will always break out after USB cable disconnect.
    4. Adjust priority of wlan main thread to 10.


2009 04 17
    1. Fiex RTSP server's bug that server will crash when quicktime teardown normally.
    2. Reduce delay time to audio test.
    3. Merged from W99803 branch:
       a. Fixed memory leak in sending mail.

2009 04 15
    1. Disable auth check for USB console tool.
    2. Update verify tool:
       a. stdout/stderr was changed since that wowwee flash tool can
          not catch pipe data from stderr.
       b. disable test for suspend.
       c. support new usb interface.

2009 04 08
    1. Use new parameter "Description" for SetVNet.cgi/GetVNet.cgi
    2. Fix a pointer error in bpserver.c

2009 04 03
    1. Fixed sscanf problemm in USB console tool.
    2. Add delay time in SetVNet.cgi

2009 03 24
    1. Add channel number settings for SetVNet.cgi.

2009 03 20
    1. Changed the ActiveX version to 1.0.0.8 and clsid to
       {115B1886-2AE0-4259-9FE4-E32A5DEE5454}

2009 03 19
    1. Modify "user privilege" fields in USB tools, so that it's
       compitable with old USB tools. 
    2. Merge RTSP drop_conn patch from W99803's branch.
    3. Do few things when rebooting.

2009 03 18
    1. Fixed Activex close crash bug.
    2. fixed activex wrong url bug.
    3. updated wbstreamplayer to support mp4 width > 640 height>48.

2009 03 16
    1. Do not disable/enable GPIO13 while accessing SPI flash,
       which may fix the problem of flash corrupted.
    2. Add -f output_file for USB command line tool.

2009 03 13
    1.updated activex for MSI filter.

2009 03 06
    1. Synchronzie with W99803's branch.

2009 03 05
    1. Add username/password for usb console tool.

2009 03 04
    1. Add new command "usb netstate", which reports:
       WIFI state:
           NO_INIT
           TRYING
           OK
       IP state:
           NO_INIT
           DHCP_TRYING
           DHCP_FAILED
           DHCP_OK
           STATIC_IP_TRYING
           STATIC_IP_FAILED
           STATIC_IP_OK
    2. "usb user" reports:
           privilege:username
    3. If a key is used, "usb wlan" reports the key with "*",
       otherwise report the blank string.
    4. Remove password for user delete.

2009 02 27
    1. Set mp4 max width to 1600 ,max height to 1200

2009 02 25
    1. update Activex for MSI facial recognition function.
2009 02 24
    1. Optimize getting mp4/jpeg and audio. 

2009 02 23
    1. Synchronzie with W99803's branch.

2009 02 20
    1. Adjust stdout/stderr in console usb tool.

2009 02 19
    1. Update console usb tool, added scan_wlan, auth enable/disable,
       updated command wlan.

2009 02 18
    1. Use large packet size for H.263.
       ( Bug not fixed for H.263 packetize. )
    2. Add current IP information to GetIP.cgi.

2009 02 13
    1. Clear error state after MCU is restored.
    2. Use timestamp on capture point for audio and video.

2009 02 11
    1. updated input-oss to check frame heap before getting audio.

2009 02 09
    1. Import source of updatepmp and vcom.
    2. Fixed the problem about setting WLAN static IP may
       cause long time waiting.

2009 02 03
    1. update wbstreamplayer for rtsp authentication.
    2. Add save user/pass for ActiveX.

2009 01 22
    1. Add authorization dialog for ActiveX.
       Have not applied JF's update for WBStreamPlayer.

2009 01 21
    1. Add basic auth for RTSP.

2009 01 20
    1. Integrate RTSP authorization with HTTP.
       Enable authorization for RTSP over UDP & TCP.

2009 01 15
    1. Test RTSP over HTTP, OK.
    2. Disable LED low battery on charge. 

2009 01 12
    1. Enlarge VCOM buffer size to 4K.

2009 01 08
    1. Fixed LED error.
    2. Update to libNS 5.13.


Host/libeCos/src/net/common/v2_0/src/getserv.c
//==========================================================================
//
//      lib/getserv.c
//
//      getservbyname(), getservbyport()
//
//==========================================================================
//####BSDCOPYRIGHTBEGIN####
//
// -------------------------------------------
//
// Portions of this software may have been derived from OpenBSD or other sources,
// and are covered by the appropriate copyright disclaimers included herein.
//
// -------------------------------------------
//
//####BSDCOPYRIGHTEND####
//==========================================================================
//#####DESCRIPTIONBEGIN####
//
// Author(s):    gthomas
// Contributors: gthomas
// Date:         2000-01-10
// Purpose:      
// Description:  
//              
//
//####DESCRIPTIONEND####
//
//==========================================================================


#include "sys/param.h"
#include "netdb.h"
#include "errno.h"

// These must return the port in network byte order.
//
// This means treated as a short, because it's a port, despite the types in
// the API and the struct being ints.
// 
// The argument to getservbyport() is also network byte order, so that code
// must change to flip before comparing.

static struct servent services[] = {
    { "ftp",      0,   21 , "tcp" },
    { "ftp-data", 0,   20 , "tcp" },
    { "domain",   0,   53 , "udp" },
    { "tftp",     0,   69 , "udp" },
    { "snmp",     0,  161 , "udp" },

    { NULL,       0,     0       , NULL  }
};

// Note that this contains no interlocking between clients of the
// interface; but this is completely typical of such APIs.

static struct servent *
setreturned( struct servent *p )
{
    static struct servent returned;

    returned.s_name     = p->s_name;
    returned.s_aliases  = p->s_aliases;
    returned.s_port     = htons(p->s_port); // return in net order
    returned.s_proto    = p->s_proto;
    return &returned;
}

struct servent *
getservbyname(const char *name, const char *proto)
{
    struct servent *p = services;
    while (p->s_name) {
        if ((strcmp(name, p->s_name) == 0) &&
            (strcmp(proto, p->s_proto) == 0)) {
            return setreturned(p);
        }
        p++;
    }
    errno = ENOENT;
    return (struct servent *)0;
}

struct servent *
getservbyport(const int num, const char *proto)
{
    struct servent *p = services;
    int port = ntohs(num);
    while (p->s_name) {
        if ((p->s_port == port) &&
            (strcmp(proto, p->s_proto) == 0)) {
            return setreturned(p);
        }
        p++;
    }
    errno = ENOENT;
    return (struct servent *)0;
}
